[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/miguepoloc/toma-decisiones-mcda/blob/main/02_topsis_iot_palmor.ipynb)

# TOPSIS, caso IoT/WSN Palmor

Elección de tecnología de comunicación (**LoRaWAN, GSM/GPRS, Sigfox, Zigbee**) para una red de sensores IoT/WSN de monitoreo agroclimático en Palmor, corregimiento de Ciénaga (Sierra Nevada de Santa Marta, Magdalena), zona con conectividad limitada verificada vía MinTIC. Los 4 criterios (Alcance de comunicación, Autonomía de batería, Infraestructura/cobertura comercial en Colombia, Madurez/viabilidad comercial del proveedor) salieron del panel de evidencia de la Sesión 1. Los valores técnicos de la matriz de decisión son reales, verificados vía WebSearch (datasheets SIMCom/DigiKey, The Things Network, Lauridsen et al. 2019 *Sensors*/MDPI, noticias de apagado de 2G en Colombia). Los pesos de los 4 criterios son los que AHP derivó en la Sesión 2 (Alcance 16.0%, Autonomía 25.0%, Infraestructura 48.8%, Madurez 10.2%), TOPSIS los recibe como dato de entrada en vez de derivarlos. Contenido completo en la Sesión 3 (TOPSIS/VIKOR) del curso.

In [1]:
!pip install -q pyDecision


[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip install --upgrade pip


In [2]:
import numpy as np
from pyDecision.algorithm import topsis_method

tecnologias = ["LoRaWAN", "GSM/GPRS", "Sigfox", "Zigbee"]
# Alcance (km), Autonomia (años), Infraestructura (escala 1-5), Madurez (escala 1-5)
dataset = np.array([
    [10,   8,   2, 5],   # LoRaWAN
    [10.5, 0.5, 3, 2],   # GSM/GPRS
    [40,   2,   5, 2],   # Sigfox
    [0.07, 1.5, 2, 4],   # Zigbee
])
pesos = [0.16, 0.25, 0.488, 0.102]
tipo = ["max", "max", "max", "max"]  # las 4 son de beneficio

## Cálculo y ranking

In [3]:
relative_closeness = topsis_method(dataset, pesos, tipo, graph=False)

print("Ranking final TOPSIS:")
for t, c in sorted(zip(tecnologias, relative_closeness), key=lambda x: -x[1]):
    print(f"  {t:10s} C_i={c:.3f}")

a1: 0.48
a2: 0.22
a3: 0.6
a4: 0.11
Ranking final TOPSIS:
  Sigfox     C_i=0.599
  LoRaWAN    C_i=0.477
  GSM/GPRS   C_i=0.224
  Zigbee     C_i=0.111


**Resultado esperado** (coincide con lo publicado en la Sesión 3 del curso): 1º Sigfox (C_i=0.599) · 2º LoRaWAN (C_i=0.477) · 3º GSM/GPRS (C_i=0.224) · 4º Zigbee (C_i=0.111). Reproduce el orden exacto de AHP.